sway-ir/src/optimize/memcpyopt.rs

```diff
@@ -1025,3 +1053,255 @@ fn load_store_to_memcopy(context: &mut Context, function: Function) -> Result<bo
+
+    for (inst, dst_sym, src_sym) in candidates {
+        match src_sym {
+            Symbol::Arg(_) => {
+                // TODO: Handle args (and remove this filter)
+                // They just require handling more instructions where they can be used.
+                continue;
+            }
+            Symbol::Local(local) => {
+                if local.get_initializer(context).is_some() {
+                    // TODO: If the source is a local and it has an initializer, we run into trouble
+                    // 1. If the destination (after transitive closure below) is not a local,
+                    //    we cannot initialize it with the source's initializer.
+                    // 2. If the destination is a local, but it already has an initializer (by itself
+                    //    or by another source in the chain), we cannot initialize it with this initializer.
+                    continue;
+                }
+                match src_to_dst.entry(src_sym) {
+                    std::collections::hash_map::Entry::Vacant(e) => {
+                        e.insert(dst_sym);
+                    }
+                    std::collections::hash_map::Entry::Occupied(e) => {
+                        if *e.get() != dst_sym {
+                            // src_sym is copied to two different dst_syms. We cannot optimize this.
+                            continue;
+                        }
+                    }
+                }
+                to_delete.insert(inst);
+            }
+        }
+    }
+
+    // Take a transitive closure of src_to_dst.
+    {
+        let mut changed = true;
+        let mut cycle_detected = false;
+        while changed {
+            changed = false;
+            src_to_dst.clone().iter().for_each(|(src, dst)| {
+                if let Some(next_dst) = src_to_dst.get(dst) {
+                    // Cycle detection
+                    if *next_dst == *src {
+                        cycle_detected = true;
+                        return;
+                    }
+                    src_to_dst.insert(*src, *next_dst);
+                    changed = true;
+                }
+            });
+        }
+        if cycle_detected {
+            // We cannot optimize in presence of cycles.
+            return Ok(modified);
+        }
+    }
+
+    // Gather the get_local instructions that need to be replaced.
+    let mut repl_locals = vec![];
+    for (_block, inst) in function.instruction_iter(context) {
+        match inst.get_instruction(context).unwrap() {
+            Instruction {
+                op: InstOp::GetLocal(sym),
+                ..
+            } => {
+                if let Some(dst) = src_to_dst.get(&Symbol::Local(*sym)) {
+                    repl_locals.push((inst, *dst));
+                }
+            }
+            _ => {
+                // Any access to a local begins with a GetLocal, we can ignore the rest
```

Comment:
If we decide to support `Arg`, let's put a TODO with the link to that GH issue here as well.